tinygrad/runtime/ops_qcom.py

```diff
@@ -0,0 +1,148 @@
+import os
+import os, ctypes, pathlib, re, fcntl, functools, mmap, struct, tempfile, hashlib, subprocess, time, array
+from types import SimpleNamespace
+from typing import Tuple, List, Any
+
+from tinygrad.device import HCQCompatCompiled, HCQCompatAllocator, HCQCompatAllocRes, Compiler, CompileError, BufferOptions
+import tinygrad.runtime.autogen.kgsl as kgsl
+import tinygrad.runtime.autogen.adreno as adreno
+from tinygrad.helpers import getenv, from_mv, mv_address, init_c_struct_t, to_mv, round_up, to_char_p_p, DEBUG, prod, PROFILE
+import tinygrad.runtime.autogen.libc as libc
+
+if getenv("IOCTL"): import extra.qcom_gpu_driver.opencl_ioctl # noqa: F401
+
+def data64_le(data): return (data & 0xFFFFFFFF, data >> 32)
+
+def parity(val: int): return (~0x6996 >> ((val ^ (val >> 16) ^ (val >> 8) ^ (val >> 4)) & 0xf)) & 1
+
+def pkt7_hdr(opcode: int, cnt: int):
+  return adreno.CP_TYPE7_PKT | cnt & 0x3FFF | parity(cnt) << 15 | (opcode & 0x7F) << 16 | parity(opcode) << 23
+
+def pkt4_hdr(reg: int, cnt: int):
+  return adreno.CP_TYPE4_PKT | cnt & 0x3FFF | parity(cnt) << 7 | (reg & 0x3FFFF) << 8 | parity(reg) << 27
+
+MAP_FIXED = 0x10
+class QcomDevice():
+  signals_page:Any = None
+  signals_pool: List[Any] = []
+
+  def __init__(self, device:str=""):
+    self.fd = os.open('/dev/kgsl-3d0', os.O_RDWR)
+    QcomDevice.signals_page = self._gpu_alloc(16 * 65536, map_to_cpu=True)
+    QcomDevice.signals_pool = [to_mv(self.signals_page.va_addr + off, 16).cast("Q") for off in range(0, self.signals_page.size, 16)]
+    cr = kgsl.struct_kgsl_drawctxt_create(flags=(2<<20) | 0x10 | 0x2)
+    self._ioctl(kgsl.IOCTL_KGSL_DRAWCTXT_CREATE, cr)
+    self.context_id = cr.drawctxt_id
+
+    # super().__init__(deitimeline_signals=[self._alloc_signal(), self._alloc_signal()])
+
+  def _ioctl(self, nr, arg):
+    ret = fcntl.ioctl(self.fd, (3 << 30) | (ctypes.sizeof(arg) & 0x1FFF) << 16 | 0x9 << 8 | (nr & 0xFF), arg)
+    if ret != 0: raise RuntimeError(f"ioctl returned {ret}")
+    return ret
+
+  def _gpu_alloc(self, size:int, flags:int=0, map_to_cpu=False):
+    size = round_up(size, align:=(2 << 20))
+    flags |= ((align << kgsl.KGSL_MEMALIGN_SHIFT) & kgsl.KGSL_MEMALIGN_MASK)
+
+    alloc = kgsl.struct_kgsl_gpuobj_alloc(size=size, flags=flags)
+    self._ioctl(kgsl.IOCTL_KGSL_GPUOBJ_ALLOC, alloc)
+    info = kgsl.struct_kgsl_gpuobj_info(id=alloc.id)
+    self._ioctl(kgsl.IOCTL_KGSL_GPUOBJ_INFO, info)
```

Comment:
You can think of switching this to the `info = kgsl.gpuobj_info(id=alloc.id)` style (like amd/nv) with some preprocessing of packets.